### PR TITLE
DAOS-5692 tools: Run daos command under valgrind.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add suppressions as required to silence go errors.
Resolve invalid reads.
Use container labels in dfuse and fix issues.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

File: utils/node_local_test.py
Comment (mjmac) [inline utils/node_local_test.py:889]: This needs to be removed before landing, right? I wonder if we could add a flag to NLT to have it optionally run in this mode, might be handy.

File: utils/node_local_test.py
Comment (mjmac) [inline utils/node_local_test.py:1790]: Why does this need to be disabled?

Comment (ashleypittman) [reply utils/node_local_test.py:1790]: It was showing the memory leak which meant that future tests weren't running.  I was trying to get more coverage to see if there are other items we should know about.

File: utils/node_local_test.py
Comment (ashleypittman) [reply utils/node_local_test.py:889]: This works quite well actually, the suppressions even render in Jenkins correctly.  I thought we'd added it a few months ago so I've got no problem leaving this in.

File: utils/node_local_test.py
Comment (ashleypittman) [reply utils/node_local_test.py]: Re-enabled, and I've change the logic to report but carry on running in the presence of valgrind issues from the daos command, so that a single test-run can report all issues.

File: src/client/dfs/duns.c
Comment (mchaarawi) [inline src/client/dfs/duns.c:501]: da_rel_path is freed externally so can't use the macro as it would complain

**Comment by ashleypittman** [comment]
It looks like there's one more memory leak landed overnight, it would be good to focus on getting this PR resolves/landed once the current batch have gone in.

**Comment by ashleypittman** [comment]
This looks to be passing now as the leaks have been resolved so this PR could land as-is.  I've rebased it to get a run at hardware-large but the previous run passed all tests other than that so can I request review in preparation for landing.

**Review by mjmac — APPROVED** [review]
LGTM

**Comment by mchaarawi** [comment]
im trying to see the test result, but i don't see the "details" for continuous-integration/jenkins/pr-head

**Comment by ashleypittman** [comment]
> im trying to see the test result, but i don't see the "details" for continuous-integration/jenkins/pr-head

That's because I just rebased it to get it to run on the Large cluster.  Results are at:

https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-5987/17/pipeline

**Comment by ashleypittman** [comment]
The downside to this PR is that there's a increase in the run-time of the NLT stage which is going to give us more contention on that cluster (although hardware Large will still be the bottleneck), I've got some ideas to reduce this or I'm open to ideas about how to move this around but we've already seen several valgrind errors with the new go based daos code so it would be good to get this re-enabled.

**Comment by mchaarawi** [comment]
> The downside to this PR is that there's a increase in the run-time of the NLT stage which is going to give us more contention on that cluster (although hardware Large will still be the bottleneck), I've got some ideas to reduce this or I'm open to ideas about how to move this around but we've already seen several valgrind errors with the new go based daos code so it would be good to get this re-enabled.

NLT was the bottleneck before 1 of the HW clusters went offline.

**Comment by ashleypittman** [comment]
> > The downside to this PR is that there's a increase in the run-time of the NLT stage which is going to give us more contention on that cluster (although hardware Large will still be the bottleneck), I've got some ideas to reduce this or I'm open to ideas about how to move this around but we've already seen several valgrind errors with the new go based daos code so it would be good to get this re-enabled.
> 
> NLT was the bottleneck before 1 of the HW clusters went offline.

I've been keeping an eye on it and it's trending that way but I don't think it's crossed that threshold yet, this is why I'm looking at ideas to reduce the time accordingly.

**Comment by mjmac** [comment]
> 
> 
> > > The downside to this PR is that there's a increase in the run-time of the NLT stage which is going to give us more contention on that cluster (although hardware Large will still be the bottleneck), I've got some ideas to reduce this or I'm open to ideas about how to move this around but we've already seen several valgrind errors with the new go based daos code so it would be good to get this re-enabled.
> > 
> > 
> > NLT was the bottleneck before 1 of the HW clusters went offline.
> 
> I've been keeping an eye on it and it's trending that way but I don't think it's crossed that threshold yet, this is why I'm looking at ideas to reduce the time accordingly.

If we don't use NLT for testing performance then we don't need to run it on real hardware, right? Maybe we could run most of the NLT suite on a larger number of VMs and just run the performance-sensitive part on real hardware?

**Comment by ashleypittman** [comment]
> > > > The downside to this PR is that there's a increase in the run-time of the NLT stage which is going to give us more contention on that cluster (although hardware Large will still be the bottleneck), I've got some ideas to reduce this or I'm open to ideas about how to move this around but we've already seen several valgrind errors with the new go based daos code so it would be good to get this re-enabled.
> > > 
> > > 
> > > NLT was the bottleneck before 1 of the HW clusters went offline.
> > 
> > 
> > I've been keeping an eye on it and it's trending that way but I don't think it's crossed that threshold yet, this is why I'm looking at ideas to reduce the time accordingly.
> 
> If we don't use NLT for testing performance then we don't need to run it on real hardware, right? Maybe we could run most of the NLT suite on a larger number of VMs and just run the performance-sensitive part on real hardware?

That's one of the options yes.  I see about a 50% performance drop under docker in NLT but if we're not testing performance that would be OK so one idea is to take the existing two NLT nodes, make them docker hosts and run up to 4 docker instances of NLT concurrently.

The trend for NLT to take longer is an existing problem, this PR makes it worse a increases the priority for a fix but I'd prefer not to block this PR on it.  If we do find that NLT is truly the bottleneck once Hardware Large is back on-line then we should drop the fault injection testing in preference to not landing this PR.

**Comment by brianjmurrell** [comment]
> That's one of the options yes. I see about a 50% performance drop under docker in NLT

What about VMs?  Same 50% or better or worse?

> but if we're not testing performance that would be OK

Indeed.  Taking longer is not a problem.  We can scale up docker/VM testing much more easily/efficiently/effectively than we can hardware.

> so one idea is to take the existing two NLT nodes, make them docker hosts and run up to 4 docker instances of NLT concurrently.

Why not just use the existing docker hosts we have?

**Comment by ashleypittman** [comment]
> > so one idea is to take the existing two NLT nodes, make them docker hosts and run up to 4 docker instances of NLT concurrently.
> 
> Why not just use the existing docker hosts we have?

That would be an option, but I'd be worried that at the current settings we'd overload it.  I think we use docker for build and have 64 "slots" in Jenkins speak, but each build only runs for 6 minutes and is effectively single-threaded, I don't think that 64 instances of NLT on a single node would be a good thing and given it takes longer (and it would be longer again if we overcommitted in this way) it's what we'd likely have.

NLT also generates much more data in the images than a simple build (which puts most of it's data to the host) so we'd need to look at local storage availability/pruning of images much more aggressively than we do with simple builds.

**Review by frostedcmos — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**
